### PR TITLE
[Feature] Add SRFolderMultipleGTDataset

### PR DESCRIPTION
**Pull request by ckkelvinchan**
## Motivation
Currently, MMEditing does not support training video SR models with custom datasets. Users can only trained on pre-determined datasets such as REDS and Vimeo-90K.

## Modification
This PR adds support to training recurrent networks (e.g. BasicVSR, IconVSR) on custom datasets. Changes are as follows:
1. Add `SRFolderMultipleGTDataset`, which allows users to train and test on their own datasets.
2. `SRTestMultipleGTDataset` is deprecated. All existing uses of `SRTestMultipleGTDataset` are replaced with `SRFolderMultipleGTDataset`, but we still keep `SRTestMultipleGTDataset` to avoid any unexpected compatibility issue. It can be removed in the future.

### CODE REVIEW

**Comment by innerlee** [comment]
Can the `custom` be avoided in the class name?

**Comment by ckkelvinchan** [comment]
Not sure if `SRMultipleGTDataset` would be ambiguous, as there are different multiple-gt datasets in MMEditing.

**Comment by innerlee** [comment]
`Custom` does not add any meanful info.

**Comment by ckkelvinchan** [comment]
> `Custom` does not add any meanful info.

How about `SRMultipleGTFolderDataset`? It corresponds to `SRFolderDataset` for image.

After this PR, there will be another named `SRVideoFolderDataset`, which is for sliding-window framework.

File: configs/restorers/basicvsr/basicvsr_vimeo90k_bd.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 
 # dataset settings
 train_dataset_type = 'SRVimeo90KMultipleGTDataset'
-val_dataset_type = 'SRTestMultipleGTDataset'
+val_dataset_type = 'SRCustomMultipleGTDataset'
```
Comment (Yshuo-Li) [inline configs/restorers/basicvsr/basicvsr_vimeo90k_bd.py:19]: What is `Custom`

File: mmedit/datasets/sr_custom_multiple_gt_dataset.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
+import glob
+import os.path as osp
+
+from .base_sr_dataset import BaseSRDataset
+from .registry import DATASETS
+
+
+@DATASETS.register_module()
+class SRCustomMultipleGTDataset(BaseSRDataset):
+    """Custom dataset for video super resolution for recurrent networks.
```
Comment (Yshuo-Li) [inline mmedit/datasets/sr_custom_multiple_gt_dataset.py:10]: Docstring failed to explain `Multiple`.
Besides, This sentence is a bit difficult to understand (... for video super resolution for recurrent networks.)

File: mmedit/datasets/sr_custom_multiple_gt_dataset.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    frames. Then it applies specified transforms and finally returns a dict
+    containing paired data and other information.
+
+    Args:
+        lq_folder (str | :obj:`Path`): Path to a lq folder.
+        gt_folder (str | :obj:`Path`): Path to a gt folder.
+        pipeline (list[dict | callable]): A sequence of data transformations.
+        scale (int): Upsampling scale ratio.
+        num_input_frames (None | int): The number of frames per iteration.
+            If None, the whole clip are extracted. If it is an integer
```
Comment (Yshuo-Li) [inline mmedit/datasets/sr_custom_multiple_gt_dataset.py:25]: If None, the whole clip will be extracted.

Comment (Yshuo-Li) [inline mmedit/datasets/sr_custom_multiple_gt_dataset.py:25]: please delete `an integer`.

File: mmedit/datasets/sr_custom_multiple_gt_dataset.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                num_input_frames = int(sequence_length)
+            else:
+                num_input_frames = self.num_input_frames
+            data_infos.append(
+                dict(
+                    lq_path=self.lq_folder,
+                    gt_path=self.gt_folder,
+                    key=sequence.replace(f'{self.lq_folder}/', ''),
+                    num_input_frames=num_input_frames,
+                    sequence_length=int(sequence_length)))
```
Comment (Yshuo-Li) [inline mmedit/datasets/sr_custom_multiple_gt_dataset.py:73]: `sequence_length = len(glob.glob(osp.join(sequence, '*.png')))`
`len(__obj: Sized) -> int`
No need to use `int(sequence_length)`

File: configs/restorers/basicvsr/basicvsr_vimeo90k_bd.py (hunk 1 above)
Comment (ckkelvinchan) [reply configs/restorers/basicvsr/basicvsr_vimeo90k_bd.py:19]: Will change to `SRFolderMultipleGTDataset`, maching `SRFolderDataset`.

File: mmedit/datasets/sr_custom_multiple_gt_dataset.py (hunk 2 above)
Comment (ckkelvinchan) [reply mmedit/datasets/sr_custom_multiple_gt_dataset.py:10]: It is explained in line 15: `The dataset loads several LQ (Low-Quality) frames and GT (Ground-Truth) frames.`. There are multiple GT frames.

File: configs/restorers/basicvsr/basicvsr_vimeo90k_bd.py (hunk 1 above)
Comment (Yshuo-Li) [reply configs/restorers/basicvsr/basicvsr_vimeo90k_bd.py]: Fine

**Review by Yshuo-Li — APPROVED** [review]

File: mmedit/datasets/sr_test_multiple_gt_dataset.py [hunk 5]
```diff
@@ -27,6 +28,10 @@ class SRTestMultipleGTDataset(BaseSRDataset):
     def __init__(self, lq_folder, gt_folder, pipeline, scale, test_mode=True):
         super().__init__(pipeline, scale, test_mode)
 
+        warnings.warn('"SRTestMultipleGTDataset" have been deprecated and '
+                      'will be removed in future release. Please use '
+                      '"SRFolderMultipleGTDataset" instead.')
```
Comment (innerlee) [inline mmedit/datasets/sr_test_multiple_gt_dataset.py:33]: "Details see https://github.com/open-mmlab/mmediting/pull/355"

**Pull request merged**